File: portafolio-backend/src/main/java/com/portafolio/portafoliobackend/services/PortafolioService.java
package com.portafolio.portafoliobackend.services;

import java.util.List;

import com.portafolio.portafoliobackend.dtos.TblPortafolioDTO;
import com.portafolio.portafoliobackend.models.entity.TblPortafolio;

public interface PortafolioService {
    
    // PERMITE LISTAR LOS PORTAFOLIOS
    List<TblPortafolioDTO> listarPortafolio();

    // PERMITE CREAR UN NUEVO PORTAFOLIO
    TblPortafolioDTO crearPortafolio(TblPortafolioDTO tblPortafolioDTO) throws Exception;

    // PERMITE OBTENER LOS DATOS DEL OBJETO
    TblPortafolioDTO obtenerPortafolioPorId(Long idPortafolio);

    // PERMITE MODIFICAR UN PORTAFOLIO
    public TblPortafolioDTO modificarPortafolio(TblPortafolioDTO tblPortafolioDTO, TblPortafolio tblPortafolio) throws Exception;

    // PERMITE ELIMINAR EL PORTAFOLIO
    public void eliminarPortafolio(TblPortafolio tblPortafolioActual) throws Exception;
    
    // PERMITE OBTENER LAS PROPIEDADES DE LOS PORTAFOLIOS
    public TblPortafolio findById(Long idPortafolio);
}

File: portafolio-frontend/src/app/services/portafolio.service.ts
import { HttpClient, HttpErrorResponse, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, throwError } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from 'src/environments/environment';
import { TblPortafolioDTO } from '../models/TblPortafolioDTO';

@Injectable({
  providedIn: 'root'
})
export class PortafolioService {

  private url = `${environment.HOST}/portafolios`;

  private httpHeaders: HttpHeaders;



  constructor(private httpClient: HttpClient) {
    this.httpHeaders = new HttpHeaders({ "Content-Type": "application/json" });
  }


  listarPortafolio(): Observable<TblPortafolioDTO[]> {
    const urlEndPoint = `${this.url}/listarPortafolio`;
    return this.httpClient.get<TblPortafolioDTO[]>(urlEndPoint, { headers: this.httpHeaders });
  }

  /**
   * Permite crear un centro de estudio.
   * @param tblPortafolioDTO,
   */

   crearPortafolio(tblPortafolioDTO: TblPortafolioDTO): Observable<TblPortafolioDTO> {
    const urlEndPoint = `${this.url}/crearPortafolio`;
    const clave = 'tblPortafolioDTOCreado';

    return this.httpClient.post<TblPortafolioDTO>(urlEndPoint, tblPortafolioDTO, { headers: this.httpHeaders })
      .pipe(
        map(respuesta => respuesta[clave] as TblPortafolioDTO )
      );
  }

  /**
   * Permite modificar un centro de estudio.
   * @param tblPortafolioDTO,
   */
   modificarPortafolio(tblPortafolioDTO: TblPortafolioDTO): Observable<TblPortafolioDTO> {
    const urlEndPoint = `${this.url}/modificarPortafolio`;
    const clave = 'tblPortafolioDTOModificada';
    return this.httpClient.put<TblPortafolioDTO>(urlEndPoint, tblPortafolioDTO, { headers: this.httpHeaders })
      .pipe(
        map(respuesta => respuesta[clave] as TblPortafolioDTO)
      );
  }

  /**
   * Permite obtener las configuraciones necesarias para el listado de centro de estudio.
   */
   obtenerConfiguracionesGenerales(idPortafolio: number): Observable<any> {
    const urlEndPoint = `${this.url}/obtenerConfiguracionesGenerales/${idPortafolio}`;
    return this.httpClient.get<any[]>(urlEndPoint);
  }

  /**
   * Permite analizar el error y responder un error específico para el usuario.
   * @param error que debe ser controlado.
   */
   private manejarError(error: HttpErrorResponse) {
    if (error.error instanceof ErrorEvent) {
      // Se produjo un error del lado del cliente o de la red. Manejarlo en consecuencia.
      console.error("Un error ha ocurrido:", error.error.message);
    } else {
      // El backend devolvió un código de respuesta fallido.
      // El cuerpo de respuesta puede contener pistas sobre lo que salió mal.
      console.error(
        `Código devuelto por el servidor ${error.status}, el cuerpo es: ${error.error.error}`
      );
    }
    // retornar un mensaje observable con un mensaje de error orientado al usuario.
    console.log(error);

    const claveRespuesta = "mensaje";
    const mensajeError = error.error[claveRespuesta] || error.message;

    return throwError(
      `${mensajeError}; por favor, vuelva a intentarlo más tarde.`
    );
  }
}
